### ISSUE TITLE
Tree: DeferredWidgets not rendered properly on Node other than root

### ISSUE THREAD

**Issue by emilwislowski** [primefaces/primefaces#6465]
**Describe the defect**
I am trying to create <p:tree> with <p:textEditor> inside each <p:treeNode>. The problem is that <p:textEditor> only renders correctly in the root of <p:treeView>. On the other hand, <p:inputText> renders correctly on every <p:treeNode>.

**Reproducer**
[primefaces-test.tar.gz](https://github.com/primefaces/primefaces/files/5458491/primefaces-test.tar.gz)

**Environment:**
 - PF Version: 8.0 and 7.0
 - JSF + Mojarra (2.2 and 2.3) and MyFaces (2.2 and 2.3)
 - Affected browsers: All

**To Reproduce**
1. run test app
2. expand child element of a tree

**Expected behavior**
Every tree node should render p:textEditor elements properly. p:textEditor renders only on the root.

**Example XHTML**
<h:form>
            <p:tree value="#{testView.root}" var="node">
                <p:treeNode>
                    <p:textEditor value="#{node}" />
                </p:treeNode>
            </p:tree>
        </h:form>

**Example Bean**
```java

@Named
@ViewScoped
public class TestView implements Serializable {

    TreeNode root;

    @PostConstruct  
    public void init() {
        root = new DefaultTreeNode("Root", null);
        TreeNode node0 = new DefaultTreeNode("Node 0", root);
        TreeNode node1 = new DefaultTreeNode("Node 1", root);

        TreeNode node00 = new DefaultTreeNode("Node 0.0", node0);
        TreeNode node01 = new DefaultTreeNode("Node 0.1", node0);

        TreeNode node10 = new DefaultTreeNode("Node 1.0", node1);

        node1.getChildren().add(new DefaultTreeNode("Node 1.1"));
        node00.getChildren().add(new DefaultTreeNode("Node 0.0.0"));
        node00.getChildren().add(new DefaultTreeNode("Node 0.0.1"));
        node01.getChildren().add(new DefaultTreeNode("Node 0.1.0"));
        node10.getChildren().add(new DefaultTreeNode("Node 1.0.0"));
        root.getChildren().add(new DefaultTreeNode("Node 2"));
    }

    public TreeNode getRoot() {
        return root;
    }

    public void setRoot(TreeNode root) {
        this.root = root;
    }
}

**Comment by melloware** [comment]
While this is definitely not working is this really a UI you are using with TextEditors in every Node of a treeview?

**Comment by emilwislowski** [comment]
Since it works in TreeTable, I thought it should work in the tree as well.

**Comment by melloware** [comment]
I personally don't think the Tree was designed for this type of behavior.  None of the examples in the Showcase show components in the tree: https://www.primefaces.org/showcase/ui/data/tree/filter.xhtml

I am tempted to mark this as WontFix and close it.

**Comment by stolp** [comment]
In my application I have a tree which nodes are composed of several components (inputText, several commandButton). This is working fine. I also have TextEditors in TabViews in a dataTable. I also plan to have TextEditors in TreeTables.

So I think it is a valid assumption to expect to have them work in a TreeView as well.

**Comment by melloware** [comment]
I wonder if it has to do with TextEditor being a deferred widget? `PrimeFaces.widget.TextEditor = PrimeFaces.widget.DeferredWidget`

**Comment by melloware** [comment]
I can definitely see in your example its has all 11 CreateWidget's for TextEditor `$(function(){PrimeFaces.cw("TextEditor","widget_j_idt6_j_idt8_2_j_idt10",{id:"j_idt6:j_idt8:2:j_idt10",toolbarVisible:true});});` but only 3 are shown and they are the 3 that are visible in the DOM. That's what makes me think its Deferred widget related.

**Comment by emilwislowski** [comment]
> I can definitely see in your example its has all 11 CreateWidget's for TextEditor `$(function(){PrimeFaces.cw("TextEditor","widget_j_idt6_j_idt8_2_j_idt10",{id:"j_idt6:j_idt8:2:j_idt10",toolbarVisible:true});});` but only 3 are shown and they are the 3 that are visible in the DOM. That's what makes me think its Deferred widget related.

You may be riight. With loading tree with all nodes expanded ` treeNode.setExpanded(true);` the TextEditor renders properly.  IMO this shoudl work whenever the tree is collapsed or expanded

**Comment by melloware** [comment]
OK that is good it confirms its DeferredWidget related.  Not quite sure how to fix this one....

**Comment by stolp** [comment]
I wonder, what is the difference between having a text editor in a hidden tree node (not working) and having it in a hidden tab in a tab view (working)?

Is it, that the tree node is generated dynamically during expansion?

**Comment by melloware** [comment]
Here is the docs for DeferredWidget: https://primefaces.github.io/primefaces/jsdocs/classes/primefaces.widget.deferredwidget.html

Which does this...
```javascript
 /**
         * Call this method in the {@link init} method if you want deferred rendering support. This method checks
         * whether the container of this widget is visible and call {@link _render} only once it is.
         */
        renderDeferred: function() {
            if(this.jq.is(':visible')) {
                this._render();
                this.postRender();
            }
```

So my guess is `jq.is(:visible')` is returning false and thus never calling `render`

**Comment by melloware** [comment]
`:visible` - https://api.jquery.com/visible-selector/

> Elements are considered visible if they consume space in the document. Visible elements have a width or height that is greater than zero.

**Comment by emilwislowski** [comment]
what about collapsed TreeTable? it also has invisible nodes, but they are rendered correctly.
[primefaces-test.tar.gz](https://github.com/primefaces/primefaces/files/5466035/primefaces-test.tar.gz)

**Comment by melloware** [comment]
So in that case somehow the DOM size must be calculated is `:visible` is returning true.  Might have to debug  some in your browser the JS calls.
